lib/browser/mqtt.ts

```diff
@@ -548,26 +620,46 @@ export class MqttClientConnection extends BufferedEventEmitter {
         if (this.currentState == MqttBrowserClientState.Connected) {
             this.currentState = MqttBrowserClientState.Stopped;
             this.emit('interrupt', -1);
+
+            /* Did we intend to disconnect? If so, then emit the event */
+            if (this.desiredState == MqttBrowserClientState.Stopped) {
+                this.emit("closed");
+            }
         }
 
         /* Only try and reconnect if our desired state is connected, ie no one has called disconnect() */
         if (this.desiredState == MqttBrowserClientState.Connected) {
             const waitTime = this.get_reconnect_time_sec();
             this.reconnectTask = setTimeout(() => {
-                    /** Emit reconnect after backoff time */
-                    this.reconnect_count++;
-                    this.connection.reconnect();
-                },
+                /** Emit reconnect after backoff time */
+                this.reconnect_count++;
+                this.connection.reconnect();
+            },
                 waitTime * 1000);
         }
     }
 
     private on_disconnected = () => {
         this.emit('disconnect');
+
+        /**
+         * This shouldn't ever occur, but in THEORY it could be possible to have on_disconnected called with the intent
+         * to disconnect without on_close called first. This would properly emit 'closed' should that unlikely event occur.
+         */
+        if (this.currentState == MqttBrowserClientState.Connected && this.desiredState == MqttBrowserClientState.Stopped) {
+            let closedCallbackData = {} as OnConnectionClosedResult;
+            this.emit("closed", closedCallbackData);
+        }
     }
 
     private on_error = (error: Error) => {
         this.emit('error', new CrtError(error))
+
+        // If we were trying to connect but got an error, then it's a connection failure!
+        if (this.desiredState == MqttBrowserClientState.Connected) {
```

Comment:
I'm not sure about this.  Connection failure is emitted in reaction to on_close in the mqtt5 client.

Reply:
It's the only way you can know whether a connection failed due to a ConnACK and we listen for it to determine if the connection failed with the connection promise. MQTT.js describes it as "Emitted when the client cannot connect (i.e. connack rc != 0) or when a parsing error occurs.".

That said though, I did move the code to where we wait for the connection to possibly fail and return the failure to the connection promise, which feels more correct and a better spot to emit the event.